<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.340">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Introduction to Computational Evolutionary Biology - 5&nbsp; Analysis of gene and genome duplication</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./phylogeny_inference_ml.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./gene_genome_duplication.html"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Analysis of gene and genome duplication</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Introduction to Computational Evolutionary Biology</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro_to_R.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to R</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data_structures_genomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">R for evolutionary biology: data structures</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./phylogeny_inference_distance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Distance-based phylogeny inference</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./phylogeny_inference_ml.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Maximum likelihood-based phylogeny inference</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./gene_genome_duplication.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Analysis of gene and genome duplication</span></span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#goals-of-this-lesson" id="toc-goals-of-this-lesson" class="nav-link active" data-scroll-target="#goals-of-this-lesson"><span class="header-section-number">5.1</span> Goals of this lesson</a></li>
  <li><a href="#identifying-duplicated-gene-pairs" id="toc-identifying-duplicated-gene-pairs" class="nav-link" data-scroll-target="#identifying-duplicated-gene-pairs"><span class="header-section-number">5.2</span> Identifying duplicated gene pairs</a></li>
  <li><a href="#classifying-duplicate-pairs" id="toc-classifying-duplicate-pairs" class="nav-link" data-scroll-target="#classifying-duplicate-pairs"><span class="header-section-number">5.3</span> Classifying duplicate pairs</a></li>
  <li><a href="#assigning-genes-into-unique-modes-of-duplication" id="toc-assigning-genes-into-unique-modes-of-duplication" class="nav-link" data-scroll-target="#assigning-genes-into-unique-modes-of-duplication"><span class="header-section-number">5.4</span> Assigning genes into unique modes of duplication</a></li>
  <li><a href="#calculating-substitution-rates" id="toc-calculating-substitution-rates" class="nav-link" data-scroll-target="#calculating-substitution-rates"><span class="header-section-number">5.5</span> Calculating substitution rates</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Analysis of gene and genome duplication</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p>Gene and genome duplications are a source of raw genetic material for evolution <span class="citation" data-cites="ohno2013evolution">(<a href="#ref-ohno2013evolution" role="doc-biblioref">Ohno 2013</a>)</span>. However, whole-genome duplications (WGD) and small-scale duplications (SSD) contribute to genome evolution in different manners. In this lesson, you will learn how to identify duplicated genes derived from WGD and SSD from whole-genome sequence data.</p>
<section id="goals-of-this-lesson" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="goals-of-this-lesson"><span class="header-section-number">5.1</span> Goals of this lesson</h2>
<p>In this lesson, you will learn to:</p>
<ul>
<li>Identify duplicated gene pairs from whole-genome protein sequences.</li>
<li>Classify duplicated genes based on their mode of duplication.</li>
<li>Calculate substitution rates (<span class="math inline">\(K_a\)</span> and <span class="math inline">\(K_s\)</span>) per substitution site.</li>
<li>Identify whole-genome duplication events in Ks distributions.</li>
</ul>
</section>
<section id="identifying-duplicated-gene-pairs" class="level2" data-number="5.2">
<h2 data-number="5.2" class="anchored" data-anchor-id="identifying-duplicated-gene-pairs"><span class="header-section-number">5.2</span> Identifying duplicated gene pairs</h2>
<p>Here, we will use the example data sets that come with the {doubletrouble} package, which contain sequences and annotation for the yeast species <em>Saccharomyces cerevisiae</em> and <em>Candida glabrata</em>. The example data are:</p>
<ol type="1">
<li><p><code>yeast_seq</code>: A list of <code>AAStringSet</code> objects (length 2, <em>Scerevisiae</em> and <em>Cglabrata</em>) containing whole-genome protein sequences.</p></li>
<li><p><code>yeast_annot</code>: A <code>GRangesList</code> object (length 2, <em>Scerevisiae</em> and <em>Cglabrata</em>) containing gene annotation (i.e., gene ranges).</p></li>
<li><p><code>diamond_intra</code>: A list of data frames (length 1, <em>Scerevisiae_Scerevisiae</em>) with the output of DIAMOND similarity searches between <em>S. cerevisiae</em> against itself.</p></li>
<li><p><code>diamond_inter</code>: A list of data frames (length 1, <em>Scerevisiae_Cglabrata</em>) with the output of DIAMOND similarity searches between <em>S. cerevisiae</em> and <em>C. glabrata</em>.</p></li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(doubletrouble)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Load data</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(yeast_seq)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(yeast_annot)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(diamond_intra)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(diamond_inter)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look at the data</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(yeast_seq)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$Scerevisiae
AAStringSet object of length 6600:
       width seq                                            names               
   [1]  4910 MSQDRILLDLDVVNQRLILFNS...SELPEMLSLILRQYFTDLASS YLR106C
   [2]  4092 MCKNEARLANELIEFVAATVTG...NYERLQAKEVASSTEQLLQEM YKR054C
   [3]  3744 MSLTEQIEQFASRFRDDDATLQ...IGSAVSPRNLARTDVNFMPWF YHR099W
   [4]  3268 MVLFTRCEKARKEKLAAGYKPL...ETLRGSLLLAINEGHEGFGLA YDR457W
   [5]  3144 MLESLAANLLNRLLGSYVENFD...SLYRNIAIAVREYNKYCEAIL YLL040C
   ...   ... ...
[6596]    25 MFSLSNSQYTCQDYISDHIWKTSSH                      YOR302W
[6597]    25 MRAKWRKKRTRRLKRKRRKVRARSK                      YDL133C-A
[6598]    24 MHSNNSRQILIPHQNENMFLTELY                       YDL247W-A
[6599]    24 MLVLYRKRFSGFRFYFLSIFKYII                       YBR191W-A
[6600]    16 MLSLIFYLRFPSYIRG                               YJR151W-A

$Cglabrata
AAStringSet object of length 5293:
       width seq                                            names               
   [1]  4880 MSIQSADTVVFDLDKAFQRRDE...VELPEMLALILRQYFSDLASQ CAGL0M11616g
   [2]  4336 MYCIIRLCLLLLYMVRFAAAIV...ITFLGIKKCIILLIIVVVSIA CAGL0I10147g
   [3]  4041 MVQRNIELARYITTLLIGVCPK...NDIESKVLDDTKQLLNSIEYV CAGL0K08294g
   [4]  3743 MASADQISEYAEKLKDDQQSLA...ISASVNPRNLAKTDISFMPWF CAGL0A01914g
   [5]  3247 MVKLTRFEKLQKEKNAEYFKPF...DTLRGSLLIAINEGHEGFGLA CAGL0K06303g
   ...   ... ...
[5289]    43 MLGAPISRDTPRKTRSKTQFFQGPIVSLITEKCTYEWGNPSIN    CAGL0M02541g
[5290]    39 MLPGGPIVVLILVGLAACIIVATIIYRKWQERQRALARF        CAGL0M03305g
[5291]    39 MLPGGVILVFILVGLAACAIVAVIIYRKWQERQRSLQRF        CAGL0L08008g
[5292]    37 MINEGQLQTLVIGFGLAMVVLVVVYHAVASTMAVKRD          CAGL0C05461g
[5293]    34 MQPTIEATQKDNTQEKRDNYIVKGFFWSPDCVIA             CAGL0C01919g</code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(yeast_annot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRangesList object of length 2:
$Scerevisiae
GRanges object with 6600 ranges and 2 metadata columns:
         seqnames        ranges strand |     type     gene_id
            &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;character&gt;
     [1]        I       335-649      + |     gene     YAL069W
     [2]        I       538-792      + |     gene   YAL068W-A
     [3]        I     1807-2169      - |     gene     YAL068C
     [4]        I     2480-2707      + |     gene   YAL067W-A
     [5]        I     7235-9016      - |     gene     YAL067C
     ...      ...           ...    ... .      ...         ...
  [6596]      XVI 939922-941136      + |     gene     YPR201W
  [6597]      XVI 943032-943896      + |     gene     YPR202W
  [6598]      XVI 943880-944188      + |     gene     YPR203W
  [6599]      XVI 944603-947701      + |     gene     YPR204W
  [6600]      XVI 946856-947338      - |     gene   YPR204C-A
  -------
  seqinfo: 31 sequences from an unspecified genome; no seqlengths

$Cglabrata
GRanges object with 5293 ranges and 2 metadata columns:
                       seqnames      ranges strand |     type      gene_id
                          &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;  &lt;character&gt;
     [1] ChrA_C_glabrata_CBS138   1608-2636      - |     gene CAGL0A00105g
     [2] ChrA_C_glabrata_CBS138   2671-4809      - |     gene CAGL0A00116g
     [3] ChrA_C_glabrata_CBS138 11697-13042      + |     gene CAGL0A00132g
     [4] ChrA_C_glabrata_CBS138 14977-15886      + |     gene CAGL0A00154g
     [5] ChrA_C_glabrata_CBS138 17913-19017      - |     gene CAGL0A00165g
     ...                    ...         ...    ... .      ...          ...
  [5289] mito_C_glabrata_CBS138 13275-13421      + |     gene    CaglfMp08
  [5290] mito_C_glabrata_CBS138 13614-14396      + |     gene    CaglfMp09
  [5291] mito_C_glabrata_CBS138 14631-14861      + |     gene    CaglfMp10
  [5292] mito_C_glabrata_CBS138 15384-16067      + |     gene    CaglfMp11
  [5293] mito_C_glabrata_CBS138 16756-17565      + |     gene    CaglfMp12
  -------
  seqinfo: 31 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(diamond_intra<span class="sc">$</span>Scerevisiae_Scerevisiae)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        query          db perc_identity length mismatches gap_open qstart qend
1 Sce_YLR106C Sce_YLR106C         100.0   4910          0        0      1 4910
2 Sce_YLR106C Sce_YKR054C          22.4    420        254       19    804 1195
3 Sce_YKR054C Sce_YKR054C         100.0   4092          0        0      1 4092
4 Sce_YKR054C Sce_YLR106C          22.4    420        254       19   1823 2198
5 Sce_YHR099W Sce_YHR099W         100.0   3744          0        0      1 3744
6 Sce_YHR099W Sce_YJR066W          22.7    339        201       12   3351 3674
  tstart tend   evalue bitscore
1      1 4910 0.00e+00   9095.0
2   1823 2198 1.30e-06     53.1
3      1 4092 0.00e+00   7940.0
4    804 1195 1.09e-06     53.1
5      1 3744 0.00e+00   7334.0
6   2074 2366 6.46e-08     57.0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(diamond_inter<span class="sc">$</span>Scerevisiae_Cglabrata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        query               db perc_identity length mismatches gap_open qstart
1 Sce_YLR106C Cgl_CAGL0M11616g          52.3   4989       2183       50      2
2 Sce_YLR106C Cgl_CAGL0K08294g          23.1    347        215       12   1064
3 Sce_YKR054C Cgl_CAGL0K08294g          26.5   4114       2753       81     83
4 Sce_YKR054C Cgl_CAGL0M11616g          22.7    419        254       17   1823
5 Sce_YHR099W Cgl_CAGL0A01914g          70.2   3761       1087       17      1
6 Sce_YDR457W Cgl_CAGL0K06303g          55.5   3318       1355       39      1
  qend tstart tend   evalue bitscore
1 4909      5 4879 0.00e+00   4439.0
2 1389   1770 2085 9.10e-07     53.5
3 4089     87 4035 0.00e+00   1376.0
4 2198    803 1194 7.59e-07     53.5
5 3744      1 3743 0.00e+00   5200.0
6 3268      1 3247 0.00e+00   3302.0</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Explore the example data sets to answer the following questions:</p>
<ol type="1">
<li><p>Is the number of sequences in <code>yeast_seq</code> the same as the number of gene ranges in <code>yeast_annot</code> for <em>S. cerevisiae</em>?</p></li>
<li><p>How many genes are there in <em>S. cerevisiae</em>’s chromosome 1?</p></li>
</ol>
<div class="callout callout-style-simple callout-none no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Q1</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(yeast_seq<span class="sc">$</span>Scerevisiae)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6600</code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(yeast_annot<span class="sc">$</span>Scerevisiae)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6600</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Q2</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(GenomicRanges<span class="sc">::</span><span class="fu">seqnames</span>(yeast_annot<span class="sc">$</span>Scerevisiae))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   I   II  III   IV    V   VI  VII VIII   IX    X   XI  XII XIII  XIV   XV  XVI 
 117  456  184  836  323  139  583  321  241  398  348  578  505  435  597  511 
Mito 
  28 </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="classifying-duplicate-pairs" class="level2" data-number="5.3">
<h2 data-number="5.3" class="anchored" data-anchor-id="classifying-duplicate-pairs"><span class="header-section-number">5.3</span> Classifying duplicate pairs</h2>
<p>{doubletrouble} classifies duplicated gene pairs into the following modes: whole-genome duplication (WGD), tandem duplication (TD), proximal duplication (PD), transposed duplication (TRD, optional), and dispersed duplication (DD).</p>
<p>To find WGD-derived genes, {doubletrouble} relies on synteny information identified with {syntenet} <span class="citation" data-cites="almeida2023syntenet">(<a href="#ref-almeida2023syntenet" role="doc-biblioref">Almeida-Silva et al. 2023</a>)</span>. Thus, we first need to process the input sequences and annotation to a format that {syntenet} understands. We do that with the <code>process_input()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(syntenet)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Process input data</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>pdata <span class="ot">&lt;-</span> <span class="fu">process_input</span>(yeast_seq, yeast_annot)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect the output</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(pdata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "seq"        "annotation"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>pdata<span class="sc">$</span>seq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$Scerevisiae
AAStringSet object of length 6600:
       width seq                                            names               
   [1]  4910 MSQDRILLDLDVVNQRLILFNS...SELPEMLSLILRQYFTDLASS Sce_YLR106C
   [2]  4092 MCKNEARLANELIEFVAATVTG...NYERLQAKEVASSTEQLLQEM Sce_YKR054C
   [3]  3744 MSLTEQIEQFASRFRDDDATLQ...IGSAVSPRNLARTDVNFMPWF Sce_YHR099W
   [4]  3268 MVLFTRCEKARKEKLAAGYKPL...ETLRGSLLLAINEGHEGFGLA Sce_YDR457W
   [5]  3144 MLESLAANLLNRLLGSYVENFD...SLYRNIAIAVREYNKYCEAIL Sce_YLL040C
   ...   ... ...
[6596]    25 MFSLSNSQYTCQDYISDHIWKTSSH                      Sce_YOR302W
[6597]    25 MRAKWRKKRTRRLKRKRRKVRARSK                      Sce_YDL133C-A
[6598]    24 MHSNNSRQILIPHQNENMFLTELY                       Sce_YDL247W-A
[6599]    24 MLVLYRKRFSGFRFYFLSIFKYII                       Sce_YBR191W-A
[6600]    16 MLSLIFYLRFPSYIRG                               Sce_YJR151W-A

$Cglabrata
AAStringSet object of length 5293:
       width seq                                            names               
   [1]  4880 MSIQSADTVVFDLDKAFQRRDE...VELPEMLALILRQYFSDLASQ Cgl_CAGL0M11616g
   [2]  4336 MYCIIRLCLLLLYMVRFAAAIV...ITFLGIKKCIILLIIVVVSIA Cgl_CAGL0I10147g
   [3]  4041 MVQRNIELARYITTLLIGVCPK...NDIESKVLDDTKQLLNSIEYV Cgl_CAGL0K08294g
   [4]  3743 MASADQISEYAEKLKDDQQSLA...ISASVNPRNLAKTDISFMPWF Cgl_CAGL0A01914g
   [5]  3247 MVKLTRFEKLQKEKNAEYFKPF...DTLRGSLLIAINEGHEGFGLA Cgl_CAGL0K06303g
   ...   ... ...
[5289]    43 MLGAPISRDTPRKTRSKTQFFQGPIVSLITEKCTYEWGNPSIN    Cgl_CAGL0M02541g
[5290]    39 MLPGGPIVVLILVGLAACIIVATIIYRKWQERQRALARF        Cgl_CAGL0M03305g
[5291]    39 MLPGGVILVFILVGLAACAIVAVIIYRKWQERQRSLQRF        Cgl_CAGL0L08008g
[5292]    37 MINEGQLQTLVIGFGLAMVVLVVVYHAVASTMAVKRD          Cgl_CAGL0C05461g
[5293]    34 MQPTIEATQKDNTQEKRDNYIVKGFFWSPDCVIA             Cgl_CAGL0C01919g</code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>pdata<span class="sc">$</span>annotation</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$Scerevisiae
GRanges object with 6600 ranges and 1 metadata column:
         seqnames        ranges strand |          gene
            &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |   &lt;character&gt;
     [1]    Sce_I       335-649      * |   Sce_YAL069W
     [2]    Sce_I       538-792      * | Sce_YAL068W-A
     [3]    Sce_I     1807-2169      * |   Sce_YAL068C
     [4]    Sce_I     2480-2707      * | Sce_YAL067W-A
     [5]    Sce_I     7235-9016      * |   Sce_YAL067C
     ...      ...           ...    ... .           ...
  [6596]  Sce_XVI 939922-941136      * |   Sce_YPR201W
  [6597]  Sce_XVI 943032-943896      * |   Sce_YPR202W
  [6598]  Sce_XVI 943880-944188      * |   Sce_YPR203W
  [6599]  Sce_XVI 944603-947701      * |   Sce_YPR204W
  [6600]  Sce_XVI 946856-947338      * | Sce_YPR204C-A
  -------
  seqinfo: 17 sequences from an unspecified genome; no seqlengths

$Cglabrata
GRanges object with 5293 ranges and 1 metadata column:
                       seqnames      ranges strand |             gene
                          &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; |      &lt;character&gt;
     [1] Cgl_ChrA_C_glabrata_..   1608-2636      * | Cgl_CAGL0A00105g
     [2] Cgl_ChrA_C_glabrata_..   2671-4809      * | Cgl_CAGL0A00116g
     [3] Cgl_ChrA_C_glabrata_.. 11697-13042      * | Cgl_CAGL0A00132g
     [4] Cgl_ChrA_C_glabrata_.. 14977-15886      * | Cgl_CAGL0A00154g
     [5] Cgl_ChrA_C_glabrata_.. 17913-19017      * | Cgl_CAGL0A00165g
     ...                    ...         ...    ... .              ...
  [5289] Cgl_mito_C_glabrata_.. 13275-13421      * |    Cgl_CaglfMp08
  [5290] Cgl_mito_C_glabrata_.. 13614-14396      * |    Cgl_CaglfMp09
  [5291] Cgl_mito_C_glabrata_.. 14631-14861      * |    Cgl_CaglfMp10
  [5292] Cgl_mito_C_glabrata_.. 15384-16067      * |    Cgl_CaglfMp11
  [5293] Cgl_mito_C_glabrata_.. 16756-17565      * |    Cgl_CaglfMp12
  -------
  seqinfo: 14 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>The processed data can be used as input to <code>run_diamond()</code>, which performs similarity searches with DIAMOND to identify intra and interspecies homologous genes. Here, as we already have pre-computed DIAMOND results in <code>diamond_intra</code> and <code>diamond_inter</code>, we will not execute <code>run_diamond()</code>, but you can see its usage in the {doubletrouble}’s vignette.</p>
<p>Now, let’s classify the duplicated gene pairs of <em>S. cerevisiae</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Classify duplicated genes using the full scheme</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>c_full <span class="ot">&lt;-</span> <span class="fu">classify_gene_pairs</span>(</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">blast_list =</span> diamond_intra,</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">annotation =</span> pdata<span class="sc">$</span>annotation,</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">binary =</span> <span class="cn">FALSE</span>,</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">blast_inter =</span> diamond_inter</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspecting the output</span></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(c_full)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Scerevisiae"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(c_full<span class="sc">$</span>Scerevisiae)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         dup1        dup2 type
1 Sce_YGR032W Sce_YLR342W  WGD
2 Sce_YOR396W Sce_YPL283C  WGD
3 Sce_YJL225C Sce_YIL177C  WGD
4 Sce_YNR031C Sce_YCR073C  WGD
5 Sce_YOR326W Sce_YAL029C  WGD
6 Sce_YJL222W Sce_YIL173W  WGD</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>How many gene pairs are there per mode? Hint: you can use the <code>table()</code> function.</p>
<div class="callout callout-style-simple callout-none no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(c_full<span class="sc">$</span>Scerevisiae<span class="sc">$</span>type)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
  DD   PD   TD  TRD  WGD 
3121   80   42    3  342 </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="assigning-genes-into-unique-modes-of-duplication" class="level2" data-number="5.4">
<h2 data-number="5.4" class="anchored" data-anchor-id="assigning-genes-into-unique-modes-of-duplication"><span class="header-section-number">5.4</span> Assigning genes into unique modes of duplication</h2>
<p>Because a gene can be duplicated multiple times by different mechanisms, the same gene can appear in multiple duplicated gene pairs. However, {doubletrouble} lets you classify genes into unique modes of duplication using the following hierarchy: <em>WGD &gt; TD &gt; PD &gt; TRD &gt; DD</em>. Let’s do that with the function <code>classify_genes()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Classify genes into unique modes of duplication</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>c_genes <span class="ot">&lt;-</span> <span class="fu">classify_genes</span>(c_full)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspecting the output</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(c_genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Scerevisiae"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(c_genes<span class="sc">$</span>Scerevisiae)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            gene type
2360 Sce_YGR032W  WGD
2361 Sce_YOR396W  WGD
2362 Sce_YJL225C  WGD
2363 Sce_YNR031C  WGD
2364 Sce_YOR326W  WGD
2365 Sce_YJL222W  WGD</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Use the data frame with classified genes to answer the questions below.</p>
<ol type="1">
<li>How many duplicated genes are there for each mode?</li>
<li>What is the duplication mode of the gene “YOR326W”?</li>
<li>What is the proportion (in %) of duplicated genes in the <em>S. cerevisiae</em> genome?</li>
</ol>
<div class="callout callout-style-simple callout-none no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Q1</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(c_genes<span class="sc">$</span>Scerevisiae<span class="sc">$</span>type)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
  DD   PD   TD  TRD  WGD 
1784   70   67    6  683 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Q2</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>c_genes<span class="sc">$</span>Scerevisiae[<span class="fu">grep</span>(<span class="st">"YOR326W"</span>, c_genes<span class="sc">$</span>Scerevisiae<span class="sc">$</span>gene), ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            gene type
2364 Sce_YOR326W  WGD</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Q3</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>(<span class="fu">length</span>(c_genes<span class="sc">$</span>Scerevisiae<span class="sc">$</span>gene) <span class="sc">/</span> <span class="fu">length</span>(pdata<span class="sc">$</span>seq<span class="sc">$</span>Scerevisiae)) <span class="sc">*</span> <span class="dv">100</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 39.54545</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="calculating-substitution-rates" class="level2" data-number="5.5">
<h2 data-number="5.5" class="anchored" data-anchor-id="calculating-substitution-rates"><span class="header-section-number">5.5</span> Calculating substitution rates</h2>
<p>Now, you can use the function <code>pairs2kaks()</code> to calculate rates of nonsynonymous substitutions per nonsynonymous site (<span class="math inline">\(K_a\)</span>), synonymouys substitutions per synonymous site (<span class="math inline">\(K_s\)</span>), and their ratios (<span class="math inline">\(\frac{K_a}{K_s}\)</span>). Rates are calculated using the codon models in KaKs_Calculator 2.0 <span class="citation" data-cites="wang2010kaks_calculator">(<a href="#ref-wang2010kaks_calculator" role="doc-biblioref">Wang et al. 2010</a>)</span>, which are implemented in the Bioconductor package {MSA2dist}.</p>
<p>For that, we will need the CDS (DNA sequences) for all <em>S. cerevisiae</em> genes, which are stored in {doubletrouble}’s example data set <code>cds_scerevisiae</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load data</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(cds_scerevisiae)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(cds_scerevisiae)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DNAStringSet object of length 6:
    width seq                                               names               
[1]  5688 ATGTCCTACAACGATCCAAACTT...AGACCTGGTCTACGATCAAATGA YGR032W
[2]  5631 ATGAACACTGATCAACAACCTTA...AGACCTGGTCAACTATAAAATAA YLR342W
[3]  5580 ATGGAAATTGAAAACGAACAGAT...CGAGCAGAGAAGTTGGAGAGTGA YPL283C
[4]  5391 ATGAAAGTTTCCGATAGGCGTAA...CGAGCAGAGAAGTTGGAGAGTGA YOR396W
[5]  5277 ATGAAAGTTTCCGATAGGCGTAA...CGAGCAGAGAAGTTGGAGAGTGA YJL225C
[6]  5277 ATGAAAGTTTCCGATAGGCGTAA...CGAGCAGAGAAGTTGGAGAGTGA YIL177C</code></pre>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Store CDS in a list</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>cds_list <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">Scerevisiae =</span> cds_scerevisiae)</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate Ka, Ks, and Ka/Ks for the first 5 gene pairs for demonstration only</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>selected_pairs <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">Scerevisiae =</span> c_full<span class="sc">$</span>Scerevisiae[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ])</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>kaks <span class="ot">&lt;-</span> <span class="fu">pairs2kaks</span>(selected_pairs, cds_list)</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect the output</span></span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>kaks</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$Scerevisiae
     dup1    dup2          Ka         Ks     Ka_Ks type
1 YGR032W YLR342W 0.058810200 5.23576000 0.0112324  WGD
2 YOR396W YPL283C 0.004007600 0.00991618 0.4041470  WGD
3 YJL225C YIL177C 0.000253203 0.00075777 0.3341420  WGD
4 YNR031C YCR073C 0.364041000 5.07223000 0.0717714  WGD
5 YOR326W YAL029C 0.395973000 5.15058000 0.0768793  WGD</code></pre>
</div>
</div>
<p>Peaks in <span class="math inline">\(K_s\)</span> distributions indicate WGD events. However, because of saturation at high <span class="math inline">\(K_s\)</span> values, only recent WGD events can be reliably identified from <span class="math inline">\(K_s\)</span> distributions. Here, to demonstrate how to identify such peaks, we will use pre-computed <span class="math inline">\(K_s\)</span> values for all soybean paralogous pairs, which is also part of the example data in {doubletrouble}.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load data</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(gmax_ks)</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(gmax_ks)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              dup1            dup2        Ks
1  GLYMA_01G000100 GLYMA_15G276800 0.3749860
3  GLYMA_01G000400 GLYMA_02G312300 0.0452761
8  GLYMA_01G000600 GLYMA_14G000400 0.1040900
10 GLYMA_01G000800 GLYMA_03G188200 0.4732800
11 GLYMA_01G000800 GLYMA_05G115300 1.2292300
12 GLYMA_01G000800 GLYMA_17G191100 1.6607100</code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Find peaks</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>peaks <span class="ot">&lt;-</span> <span class="fu">find_ks_peaks</span>(gmax_ks<span class="sc">$</span>Ks, <span class="at">npeaks =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>), <span class="at">verbose =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Optimal number of peaks: 3</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Bayesian Information Criterion (BIC): 
          E         V
2 -86950.21 -67759.55
3 -86977.96 -54203.15

Top 3 models based on the BIC criterion: 
      V,3       V,2       E,2 
-54203.15 -67759.55 -86950.21 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect output</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(peaks)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "mean"   "sd"     "lambda" "ks"    </code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(peaks)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 4
 $ mean  : Named num [1:3] 0.118 0.531 1.481
  ..- attr(*, "names")= chr [1:3] "1" "2" "3"
 $ sd    : num [1:3] 0.0549 0.2442 0.3136
 $ lambda: num [1:3] 0.352 0.433 0.215
 $ ks    : num [1:71231] 0.375 0.0453 0.1041 0.4733 1.2292 ...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize peaks</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_ks_peaks</span>(peaks)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="gene_genome_duplication_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Finally, since GMMs can lead to overfitting and overclustering in some cases, let’s assess the significance of the identified peaks with SiZer maps.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>feature<span class="sc">::</span><span class="fu">SiZer</span>(gmax_ks<span class="sc">$</span>Ks)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: no DISPLAY variable so Tk is not available</code></pre>
</div>
<div class="cell-output-display">
<p><img src="gene_genome_duplication_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We can see that the 3 peaks are indeed supported, as indicated by the blue regions.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercises
</div>
</div>
<div class="callout-body-container callout-body">
<p>Use the output of <code>find_ks_peaks()</code> with the soybean data set to answer the following questions:</p>
<ol type="1">
<li>What are the mean and standard deviation for the peaks?</li>
<li>How many gene pairs have Ks values between 0 and 0.4? What does that mean about the age of such gene pairs?</li>
</ol>
<div class="callout callout-style-simple callout-none no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Q1</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>peaks<span class="sc">$</span>mean</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        1         2         3 
0.1177495 0.5308772 1.4811982 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>peaks<span class="sc">$</span>sd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.05488597 0.24418649 0.31357988</code></pre>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Q2</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(peaks<span class="sc">$</span>ks[peaks<span class="sc">$</span>ks <span class="sc">&lt;</span> <span class="fl">0.4</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 35269</code></pre>
</div>
</div>
<p>Genes with <span class="math inline">\(K_s\)</span> between 0 and 0.4 are young, and those in this range that are present in syntenic regions have likely originated from a recent WGD event.</p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="references" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="references">References</h2>


<div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-almeida2023syntenet" class="csl-entry" role="listitem">
Almeida-Silva, Fabricio, Tao Zhao, Kristian K Ullrich, M Eric Schranz, and Yves Van de Peer. 2023. <span>“Syntenet: An r/Bioconductor Package for the Inference and Analysis of Synteny Networks.”</span> <em>Bioinformatics</em> 39 (1): btac806.
</div>
<div id="ref-ohno2013evolution" class="csl-entry" role="listitem">
Ohno, Susumu. 2013. <em>Evolution by Gene Duplication</em>. Springer Science &amp; Business Media.
</div>
<div id="ref-wang2010kaks_calculator" class="csl-entry" role="listitem">
Wang, Dapeng, Yubin Zhang, Zhang Zhang, Jiang Zhu, and Jun Yu. 2010. <span>“KaKs_calculator 2.0: A Toolkit Incorporating Gamma-Series Methods and Sliding Window Strategies.”</span> <em>Genomics, Proteomics &amp; Bioinformatics</em> 8 (1): 77–80.
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./phylogeny_inference_ml.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Maximum likelihood-based phylogeny inference</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav>
</div> <!-- /content -->



</body></html>