
# Multiple Sequence Alignment (MSA) ====
library(msa)
library(Biostrings)
fasta_path <- system.file("examples", "HemoglobinAA.fasta", package="msa")








# Calculating distance matrices =====
library(MSA2dist)
library(seqinr)









# Inferring trees ====
library(ape)









## Exercises ====

#' Use the `dist_matrix1` object to infer a phylogeny using the upgma() 
#' function from the phangorn package. Is the topology the same as the 
#' topology of `tree1`?












